### ISSUE TITLE
readAsText function causes encoding error reading large files

### ISSUE THREAD

**Issue by chauthai** [apache/cordova-plugin-file#284]
# cordova-file-plugin encoding error bug report
## Abstract
The `readAsText` function of the cordova-file-plugin causes an encoding error when reading an UTF-8 file larger than 2 megabytes on platforms iOS, Android and Windows.

## Description
The error is caused by the native implementation of the [`readAsText`](https://github.com/apache/cordova-plugin-file/blob/d135cd0e006736ea60c9dec46590de0927d57b6b/src/ios/CDVFile.m#L866) function of the `cordova-file-plugin`. Files are cut into chunks at a predefined [size](https://github.com/apache/cordova-plugin-file/blob/d135cd0e006736ea60c9dec46590de0927d57b6b/www/FileReader.js#L52) and are immediately converted to UTF-8. UTF-8 encoded characters use 4 bytes and if the cut is exactly in one character with 4 bytes, an encoding error is thrown.
You can see the UTF-8 leading byte `e2` at the end of [data-chunk.dump file](./data-chunk.dump). The missing bytes are in the next chunk and if you convert the dump file back to text a filler character is shown (e.g. a `?`).

## Workaround
We implemented a workaround using the [`readAsArrayBuffer`](https://github.com/chauthai/cordova-file-plugin-bug-report/blob/628b8cd65b295042ee953d6241d8d84fe5e27565/index-array-buffer.html#L39) function of the `cordova-file-plugin` and converting the typed array to UTF-8 in JavaScript with [TextDecoder.decode()](https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder/decode).

## How to reproduce the `encoding error`
1. Check out [sampe-app](https://github.com/chauthai/cordova-file-plugin-bug-report)
1. Install dependencies
`npm i`
1. Initialize Cordova
`npm run init`
1. Deploy on platform
`npm run ios/android/windows`
1. Press `Fire Parser` in Demo App
1. Watch the logs
1. Set a breakpoint at the throw of [encoding error](https://github.com/apache/cordova-plugin-file/blob/d135cd0e006736ea60c9dec46590de0927d57b6b/src/ios/CDVFile.m#L872) in iOS
1. Look at the `data` variable

Related to #238

**Comment by chauthai** [comment]
@janpio Any updates on this issue? The bug also occurs on the UWP platform.

**Comment by NateEag** [comment]
Just wanted to add that I ran into this problem in a project at my job.

It's not that hard to encounter it - loading a file with more than 262,144 bytes that's primarily composed of multibyte characters (that is, non-ASCII ones, in UTF-8 encoding) should do it often.

Here is a bit of interactive Python 3 that generates a UTF-8 file that should consistently break the `readAsText()` function:

```python3
>>> longstring = 'a' * 262143 + u'‘abc'
>>> file = open('testfile.txt', 'w')
>>> file.write(longstring)
>>> file.close()
```

**Comment by nip3o** [comment]
This happened to me today as well, but only on iOS. Looking at the implementation I don't see how the same issue could occur on Android however. Looking at [FileUtils.java#L1086](https://github.com/apache/cordova-plugin-file/blob/master/src/android/FileUtils.java#L1086) it seems like the whole result is written to the buffer before it gets converted into a string.

**Comment by WilsonFpz** [comment]
Same issue here.
Happened randomly in IOS only.
Please help fix it. thanks.

**Comment by YvesAmmann** [comment]
+1

**Comment by c0dem4ster** [comment]
+1

**Comment by NateEag** [comment]
For the +1s, some details would be nice. 

In my case, the workaround suggested in the report worked just fine.

Have you tried it? If so, did it solve your problem?

**Comment by YvesAmmann** [comment]
@NateEag: Yes, the workaround with `readAsArrayBuffer` did worked for me – but it’s only a workaround, not a fix. :)

**Comment by NateEag** [comment]
Gotcha. Yes, a real fix would obviously be better. I just wanted to be sure there wasn't some subtlety being missed that meant the workaround wasn't fully functional. 

Thanks for the details, @YvesAmmann !

**Comment by MauroIT** [comment]
Anyone found a workaround for the Windows platform?

**Comment by NateEag** [comment]
@MauroIT Does the workaround suggested above not work on Windows? If so, how does it fail?

**Comment by MauroIT** [comment]
I've added a Polyfill for [window.TextDecoder](https://github.com/anonyco/FastestSmallestTextEncoderDecoder). It works on Windows now :)

**Comment by ouou12138** [comment]
have you try with 30m+ file? I got an error like this when i try a 30m+ large file：
```json
{"type":"error","bubbles":false,"cancelBubble":false,"cancelable":false,"lengthComputable":false,"loaded":0,"total":0,"target":{"_readyState":2,"_error":{"code":1},"_result":null,"_progress":0,"_localURL":"http://localhost/__cdvfile_sdcard__/Download/yuanshen_4.0.0.apk","_realReader":{}}}
```
